memgpt/server/server.py

```diff
@@ -51,8 +54,50 @@ def run_command(self, user_id: str, agent_id: str, command: str) -> Union[str, N
         raise NotImplementedError
 
 
+class LockingServer(Server):
+    """Basic support for concurrency protections (all requests that modify an agent lock the agent until the operation is complete)"""
+
+    # Locks for each agent
+    _agent_locks = {}
+
+    @staticmethod
+    def agent_lock_decorator(func: Callable) -> Callable:
+        @wraps(func)
+        def wrapper(self, user_id: str, agent_id: str, *args, **kwargs):
+            # print("Locking check")
+
+            # Initialize the lock for the agent_id if it doesn't exist
+            if agent_id not in self._agent_locks:
+                # print(f"Creating lock for agent_id = {agent_id}")
+                self._agent_locks[agent_id] = Lock()
```

Comment:
@sarahwooders FYI:

- This locks on the `agent_id`, but it should probably lock on `(user_id, agent_id)`?
- This locking mechanism assumes that there is only one server process/object running at once (which I think is a fine assumption, but I guess we could add locking files if we wanted to support many copies of server processes running at once all using the same data?)
- Did not do much reasoning about deadlocks other than releasing the lock via a 'finally' (could also add timeouts etc)